\documentclass{article}

\usepackage{natbib}
\usepackage{graphicx}
\usepackage{amsmath}
\usepackage{indentfirst}
\usepackage{url}
\usepackage[utf8]{inputenc}

\DeclareMathOperator{\var}{var}
\DeclareMathOperator{\cov}{cov}

% \VignetteIndexEntry{msap user's guide}

\begin{document}
\SweaveOpts{concordance=TRUE}

<<foo,include=FALSE,echo=FALSE>>=
options(keep.source = TRUE, width = 60)
foo <- packageDescription("msap")
@

\title{\emph{msap} (v. \Sexpr{foo$Version}) - User's Guide }
\author{Andres Perez-Figueroa}
\maketitle

\section{Introduction}
\emph{msap} provides a deep analysis of epigenetic variation starting from a binary data matrix indicating the presence or absence of EcoRI-HpaII and EcoRI-MspI fragments, typical of MSAP technique. After comparing the data from both enzyme combinations, the program determines if each fragment is susceptible to methylation (representative of epigenetic variation) or if there is no evidence of methylation (representative of genetic variation). Different analyses of the variation and differentiation (genetic and epigenetic) among user-defined groups of samples are then performed, as well as the classification of the methylation occurrences in those groups. A comprehensive report of the analyses and several useful plots could help researchers to asses the epigenetic variation in their experiments using MSAP. Standard AFLP data is also suitable to be analyzed. All analyses follow a band-based strategy (\citealp{Bonin07}). There are several examples in the literature of MSAP experiments using some of the analyses provided by \emph{msap} (see \citealp{Herrera10, Moran11, Chwedorzewska12, Gupta12, Rodriguez12}) so the package could be useful in those kind of approaches.

The package is intended to be easy to use even for those people non-familiar to the R environment. Advanced users could take advantage of available source code to adapt \emph{msap} for more complex analyses.

\section{R basics. All you need to know about R to run \emph{msap}}

The only knowledge required for installing and running \emph{msap} is about how to open an R session in your computer. R is a statistical programming language that provides many built-in functions for performing statistical analysis and is also flexible to allow users to write their own functions.

R can be downloaded and installed for free from the website http://cran.r-project.org where detailed instructions for installing R on any operating system are provided. Accessing R is different for every operating system. For windows users, simply double click the R icon that is created after installation. For Mac users, you can double click the R icon under your Applications menu. On Linux, in the terminal window simply type R at the command prompt and R will be opened within the terminal window.

When you open R, no matter the operating system you are using, you will see the command prompt
symbol \verb@>@ which simply means that R is waiting for you to give it a command. To quit R, simply type \verb@q()@ in the command prompt and R will ask you if you want to save the workspace before quitting.
And that's all.

\section{Installing \emph{msap}}
You can install \emph{msap} automatically from a R session.
To install the last stable version from CRAN:
\begin{verbatim}
> install.packages("msap")
\end{verbatim}
To get the last daily development version from R-Forge:
\begin{verbatim}
>  install.packages("msap", repos= c("http://R-Forge.R-project.org", getOption("repos")))
\end{verbatim}

The above instructions should install \emph{msap} and all required dependecies.


\section{Preparation of data}
In order to use \emph{msap} for analyzing your results from a MSAP experiment, you need to provide a data file with a binary matrix (1/0) indicating the presence or absence of EcoRI-HpaII and EcoRI-MspI fragments in a bunch of samples of two or more populations/groups.
Data file should be a .csv file with markers as columns and two rows by sample, one for each isoschizomer reaction. File could be edited in the a spreadsheet of your choice (see Figure~\ref{fig:data}) and then saved as csv (with ',' as field separator). The final text file should look like Figure~\ref{fig:csv} if opened in a text editor.

The first row should include the markers name/references, these should be ordered by primer combination (if applicable) as analysis will be seprated for them. The first column should provide the label for the group where the sample is included, with the aim to make comparisons between different groups. Second column is reserved for an arbitrary label (i.e. to name the sample). Third column should identify the isoschizomer with 'HPA' or 'MSP'. From version 1.1.0 the fragment classification is done within the different primer combinations using their own error rate. Marker list (columns fourth to end) should be ordered by primer combinations (take note of the number of markers per primer combination, as they will be required as argument later).

If you want to analyze a standard AFLP dataset (or any other dominant markers coded by a 1/0 matrix) the datafile format is the same, but the program will ignore content of third column and treat all rows as independent samples.


\begin{figure}
\centering
\includegraphics[width=1\textwidth]{datafile.png}
\caption{Data format as seen in a spreadsheet for edition}
\label{fig:data}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=1\textwidth]{csvfile.png}
\caption{Final data format in the .csv file}
\label{fig:csv}
\end{figure}

\section{Running \emph{msap}}
We start by loading the msap package into an R session\footnote{Windows users have a step-by-step guide in section \ref{sec:WinRun}}.
<<>>=
library(msap)
@

It is highly recommended to change the working directory to that where datafile is located. Windows users can use the menu item 'File>Change dir' and choose the appropiate folder. To change the working directory within an R console run the command \verb@setwd(dir)@ where \emph{dir} is the absolute path to the directory. The output files created by \emph{msap} will be saved in that working directory.

Once we are in the right working directory with an appropiate data file, we can run all analyses of \emph{msap} with a single command (change "example.csv" by the name of your datafile, keeping the quotes, and change "Example" by a custom name, keeping quotes, to identify your data):

\begin{verbatim}
> msap("example.csv", name = "Example")
\end{verbatim}

Those users familiar with R would prefer to store the returning list (with several useful data for futher analysis) of msap:

\begin{verbatim}
> myList <- msap("example.csv", name = "Example")
\end{verbatim}

On execution, \emph{msap} will run all analyses and will show an on screen text report with the results:


<<echo=FALSE>>=
res <- msap("example.csv",name="Example")
@

From version 1.1.2, msap returns a list with data useful for further analysis. These are the data slots stored in the list:
\begin{itemize}
    \item{groups}{A factor with the name of the group of every individual analysed}
    \item{patterns}{A list showing the MSAP patterns (11, 10, 01 and 00 coded as u, h, i, f) in all groups }
    \item{transformed.MSL}{A data frame including the binary (1: unmmethylated, 2: methylated) values for those loci classified as MSL}
    \item{transformed.NML}{A data frame including the binary (1: unmmethylated, 2: methylated) values for those loci classified as NML}
    \item{DM.MSL}{A distance matrix object between all individuals for those polymorphic loci classified as MSL}
    \item{DM.NML}{A distance matrix object between all individuals for those polymorphic loci classified as NML}
    \item{DM.AFLP}{A distance matrix object between all individuals when analysing AFLP data}
\end{itemize}

In addition, \emph{msap} also produces some exploratory figures that are directly saved into .png files:

\begin{itemize}
\item A plot of Principal Coordinate Analysis (PCoA), see Figure~\ref{fig:msl}, showing the first two axis. They are saved as '\emph{<name>}-MSL.png' and '\emph{<name>}-NML.png' for MSL and NML respectively, where \emph{<name>} represents the name passed as argument in the calling to msap function.

\item A neighbor-joining tree of all samples (see Figure~\ref{fig:tree}) saved as '\emph{<name>}-MSL-NJ.png' and '\emph{<name>}-NML-NJ.png' for MSL and NML respectively

\end{itemize}


\begin{figure}
\centering
\includegraphics[width=0.7\textwidth]{Example-MSL.png}
\caption{Representation of Principal Coordinate Analysis (PCoA) for epigenetic (MSL) differentiation between groups. The first two coordinates (C1 and C2) are shown with the percentage of variance explained by them. Different point types represent individuals from different groups. Group labels show the centroid for the points cloud in each group. Ellipses represent the average dispersion of those poins around their centre. The long axis of the ellipse shows the direction of maximum dispersion and the short axis, the direction of minimum dispersion.}
\label{fig:msl}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=0.7\textwidth]{Example-MSL-NJ.png}
\caption{Neighbor-Joining tree of all samples (numbered labels at the tips) for epigenetic (MSL) distances. Colors represent different groups/populations.}
\label{fig:tree}
\end{figure}

Furthermore, some files (.csv format) with useful data are produced, in the working directory, to be processed with external programs: 

\begin{itemize}
\item A couple of .csv files with the transformed data matrices for MSL and NML. They are saved as '\emph{<name>}-MSL-transformed.csv' and '\emph{<name>}-NML-transformed.csv' for MSL and NML respectively, where \emph{<name>} represents the name passed as argument in the calling to msap function.

\item A table with the PCoA coordinates in all axis for each sample. It is saved as '\emph{<name>}-MSL-PCoA.coor.csv' and '\emph{<name>}-NML-PCoA.coor.csv' for MSL and NML respectively.

\item A table with the eigenvalues for all axis obtained by PCoA saved as '\emph{<name>}-MSL-PCoA.eige.csv' and '\emph{<name>}-NML-PCoA.eige.csv' for MSL and NML respectively

\end{itemize}



\subsection{Further options}
In the previous section, the basic use of \emph{msap} was described. However, it is possible to set some different options in the program if passed as arguments to the msap() function. 

Here is the full usage of msap() function including all the arguments and their default values (if applicable). Except for the 'datafile' that is required, all other arguments are optional.
\begin{verbatim}
msap(datafile, name=datafile, no.bands = "u", nDec=4, 
meth=TRUE, rm.redundant=TRUE, rm.monomorphic=TRUE, 
do.pcoa=TRUE, do.shannon=TRUE, do.amova=TRUE, 
do.pairwisePhiST=TRUE, do.cluster=TRUE, use.groups=NULL, 
do.mantel=FALSE, np.mantel=1000, loci.per.primer=NULL, 
error.rate.primer=NULL,  uninformative=TRUE)
\end{verbatim}

\begin{description}

  	\item[datafile]{
  		String containing the url of the csv file with the data. Required.
	}
  	\item[name]{
  		a name for the dataset to be included in the output files. By default, the name of the given datafile is used.  
	}
  	\item[no.bands]{
  		String to indicate how to deal with HPA-/MSP- pattern in MSL, with two possible values: 'h' for hypermethylated state and 'u' for uninformative state (default). 
  		
  		With 'h' researcher assumes that HPA-/MSP- (no band for both isoschizomers) pattern represents full methylation of cytosines in the target (hypermethylation) ignoring the chance of genetic change in the target. This approach could be very useful when no much genetic differences are expected between samples (i.e. the same individuals are compared for different treatments). In this case HPA-/MSP- will compute as 'methylated' state in the binary matrix used for subsequent analysis. See \citealp{Gupta12} for an example of this kind of approach.
  		
  		With 'u' (default value), researcher considers that pattern as uninformative as could be caused by a missing target (mutation) as well as by hypermethylation. So, in this case, HPA-/MSP- will compute as missing values in the binary matrix used for subsequent analysis. This should be the chosen approach in almost all ecology studies. See \citealp{Herrera10} for an example.
	}
  
  	\item[nDec]{
  		number of digits of precision for floating point output.
	}
	\item[meth]{
  	Logical value switching between MSAP ('TRUE') and standard AFLP ('FALSE') analysis. The difference lies in that for AFLP (or any other dominant marker coded by a 1/0 matrix) the 'enzyme' column is ignored and every row in data represent an independent sample, without combination of data. This is useful to compare MSAP (epigentic) and AFLP (genetic) differentiation using the same analyses.
	}
	\item[rm.redundant]{
		Not implemented yet.
	}
	\item[rm.monomorphic]{
		Logical value switching between the removal ('TRUE', by default) of monomorphic fragments (defined as those with only one state or just one ocurrence of the second state across the whole dataset) after data transformation.
	}
	\item[do.pcoa]{
		Option switcher for doing a Principal Coordinate Analysis for variation between groups. TRUE by default.
	}
	\item[do.shannon]{
		Option switcher for Shannon's Diversity Index comparison between MSL and NML.
	}
	\item[do.amova]{
		Option switcher for doing an AMOVA for differentiation between groups. TRUE by default.
	}
	\item[do.pairwisePhiST]{
		Logical value switching between the calculation of the pairwise $\Phi_{st}$ between pairs of groups/populations ('TRUE' by default) or skip it ('FALSE').
	}
	\item[do.cluster]{
		Calculates and plots a Neighbour-Joining tree ('TRUE' by default) or skip it ('FALSE').
	}
	\item[do.mantel]{
		Performs a Mantel test to obtain correlation between MSL and NML ('TRUE') or skip it ('FALSE' by default).
	}
	\item[np.mantel]{
		Gives the number of permutations for the above Mantel test (1000 by default) or skip it ('FALSE').
	}
	\item[use.groups]{
		Gives the groups/populations/treatments of the datafile to be analysed. By default all groups are considered for the analysis. To provide a subset of the groups a vector should be passed with the names of groups to be included. For example, in a datafile with 5 groups (Control, pop1, pop2, pop3 and pop4) we are interested only in Control and pops 1 and 3. Then, msap should be called with 'use.groups=c('Control','pop1','pop3')'.
	}
	\item[loci.per.primer]{
		Vector providing the number of loci/fragements obtained per primer combination. Fragment classsification is performed independenty for each primer combination. These fragment should be ordered in the datafilein the sameway as specified here. If this is nor provided (by default) then all fragments should be analyzed as they come from a single combination. For example, if there are three primer combinations with 135, 234 and 210 loci each, then msap should be called with 'loci.per.primer=c(135,234,210)'
	}
	\item[error.rate.primer]{
		Gives the repeatibility value of MSAP assays for each primer combination. It provides a threshold to consider methylation events as genotyping errors.
	}
	\item[uninformative]{
		This is a deprecated argument kept for compatilibility porpouses with previous versions.
	}
\end{description}

\section{Frequently Asked Questions}
These are some questions that users have made about the use of \emph{msap}. Here I try to answer them with step-by-step procedures.
\subsection{I'm very new into R and I run it under Windows, what steps must I follow to analyse my MSAP data?}
Please follow these instructions carefully:
\label{sec:WinRun}
\begin{enumerate}
\item Open R by double clicking its icon in your program list or shortcut in desktop.

\item Check the working directory by typing in the R console: \emph{getwd()}
      This would return the current working directory (i.e: \emph{[1] "C:/Documents and Settings/andres/Mis documentos"}), if this is not the desired (i.e. that with the input file) then you have to change it (or, alternatively, move your datafile to that folder)

\item Change the working directory. You have two alternatives to do this:
	\begin{itemize}
    \item By typing: \emph{setwd("C:/Documents and Settings/andres/Mis documentos/TheFolderWithMyData/")}   changing the route to that folder where your datafile lies.
    \item By selecting the folder from the R menu item File > Change dir.  The selected folder (that containing your datafile) will be the working directory now.
    \end{itemize}

\item Optionally check that change was right by typing \emph{getwd()} again and check if your datafile is in the working directory by typing \emph{dir()} that will list all files in the working directory.

\item Now R is ready. Load the package (that you have installed before) by typing: \emph{library(msap)}  and run the analyses by typing: \emph{msap("yourdatafile.csv", name="YourAnalysis")}   or adding further option parameters (within the parenthesis and separated by commas) as described in the Further Options section. With some experience you would skip steps 2 and 4.


\item A text report will appear in the R window (that you can copy-paste) and some files will be created (starting by "YourAnalysis-" and described in the user guide) in the working directory/folder.
\end{enumerate}
\subsection{Can I analyze AFLP data? How?}
Yes, it is possible to analyse standard AFLP data by skipping all data transformation and classification of MSL/NML and going directly to diversity/differentiation analysis. To do this, there is just a couple of steps:
\begin{enumerate}
\item The file format is the same that for MSAP data, but for AFLP the third column (enzyme label) will be ignored AND each row is an individual sample (recall that for MSAP there were 2 rows per sample). It sould be something like that:
\begin{verbatim}
,,,m1,m2,m3,m4.....
Pop1,i1,ANYTHING,1,1,0,1.....
Pop1,i2,ANYTHING,0,0,1,1.....
Pop1,i3,ANYTHING,1,0,0,0.....
Pop2,i4,ANYTHING,0,0,1,1.....
Pop2,i6,ANYTHING,1,0,0,0.....
Pop2,i6,ANYTHING,1,1,1,1.....
\end{verbatim}

\item Run the program with the option meth=FALSE (so it skips all methylation-related stuff):
\begin{verbatim}
msap("MyAFLPdatafile.csv", name="AFLPanalysis", meth=FALSE)
\end{verbatim}
\end{enumerate}
This is useful for those experiments that combine MSAP profiles and AFLP profiles to compare epigenetic vs. genetic variation, as both can be analysed with the same program.

\subsection{What if I need another kind of analysis for my MSAP data?}
The analyses currently provided by \emph{msap} are limited but this does not mean that further analysis could not be added in the future. I try to keep the package updated to allow exploratory assays of epigenetic diversity and differentiation, focused on the field of evolutionary ecology. If you are trying to analyse your MSAP data and options in \emph{msap} do not fit your requirements, then you have three alternatives to get your tasks done by \emph{msap}:
\begin{itemize}
\item If you have programming skills or experience with R, then you can get the source code of \emph{msap} and adapt it to your needs. That is the main advantage of open source software!
\item If you have programming skills or experience with R, and want to collaborate with me to expand \emph{msap}, then do not hesitate to contact me for joining the development team.
\item If you are not familiar with R or do not feel confident to make code yourself, then use the support tracker (in msap website) to request new features in \emph{msap}. I'll try to implement them as soon as possible.
\end{itemize}

\begin{thebibliography}{9}

\bibitem[Bonin et al.(2007)]{Bonin07} Bonin, A., Ehrich, D., and Manel, S. (2007). Statistical analysis of amplified fragment length polymorphism data: a toolbox for molecular ecologists and evolutionists. \emph{Molecular ecology}, 16(18), 3737-58. 

\bibitem[Chwedorzewska and Bednarek(2012)]{Chwedorzewska12} Chwedorzewska, K. J., and Bednarek, P. T. (2012). Genetic and epigenetic variation in a cosmopolitan grass Poa annua from Antarctic and Polish populations. \emph{Polish Polar Research}, 33(1), 63-80. 

\bibitem[Gupta et al.(2012)]{Gupta12} 
Gupta, V., Bijo, J., Kumar, M., Reddy, C. R. K., and Jha, B. (2012). Detection of Epigenetic Variations in the Protoplast-Derived Germlings of Ulva reticulata Using Methylation Sensitive Amplification Polymorphism (MSAP). \emph{Marine biotechnology} DOI: 10.1007/s10126-012-9434-7

\bibitem[Herrera and Bazaga(2010)]{Herrera10} 
Herrera, C. M., and Bazaga, P. (2010). Epigenetic differentiation and relationship to adaptive genetic divergence in discrete populations of the violet Viola cazorlensis. \emph{The New phytologist}, 187(3), 867-76. 

\bibitem[Moran and Perez-Figueroa(2011)]{Moran11}Moran, P., and Perez-Figueroa, A. (2011). Methylation changes associated with early maturation stages in the Atlantic salmon. \emph{BMC genetics}, 12(1), 86. 

\bibitem[Rodriguez et al.(2012)]{Rodriguez12} 
Rodriguez, C. M., Moran, P., Lago, F., Beckmann, M., and Consuegra, S. (2012). Detection and quantification of tissue of origin in salmon and veal products using methylation sensitive AFLPs, \emph{Food Chemistry} 131, 1493-1498. 

\end{thebibliography}

\section{Session Info}
This document was created using the following:

<<sessionInfo>>=
sessionInfo()
@


\end{document}
